yarn-project/simulator/src/public/public_tx_simulator.ts

```diff
@@ -229,4 +228,130 @@ export class PublicTxSimulator {
+   * @param callRequest - The enqueued call to execute
+   * @param executionRequest - The execution request (includes args)
+   * @returns The result of execution.
+   */
+  private async simulateEnqueuedCall(
+    phase: TxExecutionPhase,
+    context: PublicTxContext,
+    callRequest: PublicCallRequest,
+    executionRequest: PublicExecutionRequest,
+  ): Promise<AvmFinalizedCallResult> {
+    const stateManager = context.state.getActiveStateManager();
+    const address = executionRequest.callContext.contractAddress;
+    const selector = executionRequest.callContext.functionSelector;
+    const fnName = await getPublicFunctionDebugName(this.worldStateDB, address, selector, executionRequest.args);
+
+    const availableGas = context.getGasLeftForPhase(phase);
+    // Gas allocated to an enqueued call can be different from the available gas
+    // if there is more gas available than the max allocation per enqueued call.
+    const allocatedGas = new Gas(
+      /*daGas=*/ availableGas.daGas,
+      /*l2Gas=*/ Math.min(availableGas.l2Gas, MAX_L2_GAS_PER_ENQUEUED_CALL),
+    );
+
+    const result = await this.simulateEnqueuedCallInternal(
+      context.state.getActiveStateManager(),
+      executionRequest,
+      allocatedGas,
+      context.getTransactionFee(phase),
+      fnName,
+    );
+
+    const gasUsed = allocatedGas.sub(result.gasLeft);
+    context.consumeGas(phase, gasUsed);
+    this.log.verbose(
+      `[AVM] Enqueued public call consumed ${gasUsed.l2Gas} L2 gas ending with ${result.gasLeft.l2Gas} L2 gas left.`,
+    );
+
+    // TODO(dbanks12): remove once AVM proves entire public tx
+    context.updateProvingRequest(
+      this.realAvmProvingRequests,
+      phase,
+      fnName,
+      stateManager,
+      executionRequest,
+      result,
+      allocatedGas,
+    );
+
+    stateManager.traceEnqueuedCall(callRequest, executionRequest.args, result.reverted);
+
+    if (result.reverted) {
+      const culprit = `${executionRequest.callContext.contractAddress}:${executionRequest.callContext.functionSelector}`;
+      context.revert(phase, result.revertReason, culprit); // throws if in setup (non-revertible) phase
+    }
+
+    return result;
+  }
+
+  /**
+   * Simulate an enqueued public call, without modifying the context (PublicTxContext).
+   * Resulting modifcations to the context can be applied by the caller.
+   *
+   * This function can be mocked for testing to skip actual AVM simulation
+   * while still simulating phases and generating a proving request.
+   *
+   * @param stateManager - The state manager for AvmSimulation
+   * @param context - The context of the currently executing public transaction portion
+   * @param executionRequest - The execution request (includes args)
+   * @param allocatedGas - The gas allocated to the enqueued call
+   * @param fnName - The name of the function
+   * @returns The result of execution.
+   */
+  private async simulateEnqueuedCallInternal(
```

Comment:
I'm not saying we do, but you could also consider making this protected and subclass for testing instead of mocking.

I'm wondering: would protected be enough for easy mocking as if it was public? hmm

Reply:
no unfortunately, protected didn't change the need for that bulky syntax when mocking